libsql-sqlite3/src/wal.c

```diff
@@ -1729,6 +1761,51 @@ static int walIteratorInit(Wal *pWal, u32 nBackfill, WalIterator **pp){
   return rc;
 }
 
+static int walIteratorRevInit(Wal *pWal, u32 nBackfill, struct WalIteratorRev *p, u32 mxSafeFrame, int getMissingFrame){
+    WalIterator *pIter;
+    u32 *frames;
+    u32 iFrame, iPageno;
+    int rc;
+
+    frames = (u32*)sqlite3MallocZero((pWal->hdr.mxFrame + 1) * sizeof(u32));
+    if (!frames) return SQLITE_NOMEM_BKPT;
+    rc = walIteratorInit(pWal, nBackfill, &pIter);
+    if (rc || !pIter) {
+        sqlite3_free(frames);
+        return rc;
+    }
+
+    while (walIteratorNext(pIter, &iPageno, &iFrame) == 0) {
```

Comment:
so do I understand it correctly that `walIteratorRevInit` is actually linear, and it iterates the whole WAL structure, remembering all the frame positions in memory, so that we can iterate it backwards later?

If it's the case, it sounds like a memory usage and performance regression, and even if we want to bet on it for libsql-server, it should be opt-in, it could kill lots of valid use cases where the WAL is large and there's not that much memory available -- esp. if somebody runs SQLite with soft and hard memory limits set up. Any attempt at checkpoint will just return an allocation failure.